app/models/registry_client.rb

```diff
@@ -12,41 +26,81 @@ def initialize(host, use_ssl = true, username = nil, password = nil)
     @password = password
   end
 
-  def credentials?
-    @username && @password
-  end
-
+  # Retrieves the manifest for the required repository:tag. If everything goes
+  # well, it will return a parsed response from the registry, otherwise it will
+  # raise either ManifestNotFoundError or a RuntimeError.
   def manifest(repository, tag = "latest")
     res = get_request("#{repository}/manifests/#{tag}")
     if res.code.to_i == 200
       JSON.parse(res.body)
     elsif res.code.to_i == 404
-      raise ManifestNotFoundError, "Cannot find manifest for #{repository}:#{tag}"
+      raise NotFoundError, "Cannot find manifest for #{repository}:#{tag}"
     else
-      raise "Something went wrong while fetching manifest for #{repository}:#{tag}:" \
-        "[#{res.code}] - #{res.body}"
+      raise "Something went wrong while fetching manifest for " \
+        "#{repository}:#{tag}:[#{res.code}] - #{res.body}"
     end
   end
 
+  # Fetches all the repositories available in the registry, with all their
+  # corresponding tags. If something goes wrong while fetching the repos from
+  # the catalog (e.g. authorization error), it will raise an exception.
+  #
+  # Returns an array of hashes which contain two keys:
+  #   - name: a string containing the name of the repository.
+  #   - tags: an array containing the available tags for the repository.
+  def catalog
+    res = get_request("_catalog")
+    if res.code.to_i == 200
+      catalog = JSON.parse(res.body)
+      add_tags(catalog["repositories"])
+    elsif res.code.to_i == 404
+      raise NotFoundError, "Could not find the catalog endpoint!"
+    else
+      raise "Something went wrong while fetching the catalog"
```

Comment:
I think it would be better to add the http code and maybe the body as part of the exception message. Just to make debugging easier (think of bug reports from customers).